lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+          auto newSubfield =
+              builder->create<SubfieldOp>(op.getLoc(), getPortType(field),
+                                          newInstance.getValue(), field.suffix);
+
+          // Use the original result and the sub-field name (without the prefix)
+          // to remember the new subfield.
+          auto fieldName = StringRef(field.suffix).drop_front(fieldname.size());
+          setBundleLowering(result, fieldName, newSubfield);
+        }
+      }
+    }
+  }
+
+  Optional<Value> newValue = getBundleLowering(rootBundle, suffix);
+  if (newValue.hasValue()) {
+    // If the lowering for the current result exists, we have finished
+    // flattening. Replace the current result with the flattened value.
+    result.replaceAllUsesWith(newValue.getValue());
+  } else {
+    // Otherwise, remember the root bundle and the suffix so far.
+    setSubfieldLowering(result, rootBundle, std::string(suffix));
+  }
+
+  removeOp(op);
+
+  return success();
+}
+
+// Lowering connects only has to deal with one special case: connecting two
+// bundles. This situation should only arise when both of the arguments are a
+// bundle that was:
+//   a) originally a block argument
+//   b) originally an instance's port
+//
+// When two such bundles are connected, none of the subfield visits have a
+// chance to lower them, so we must ensure they have the same number of
+// flattened values and flatten out this connect into multiple connects.
+LogicalResult FIRRTLTypesLowering::visitStmt(ConnectOp op) {
+  Value lhs = op.lhs();
+  Value rhs = op.rhs();
+
+  // If we aren't connecting two bundles, there is nothing to do.
+  if (!(lhs.getType().isa<BundleType>() && rhs.getType().isa<BundleType>()))
+    return success();
+
+  // Get the lowered values for each side.
+  SmallVector<Value, 8> lhsValues;
+  getAllBundleLowerings(lhs, lhsValues);
+
+  SmallVector<Value, 8> rhsValues;
+  getAllBundleLowerings(rhs, rhsValues);
+
+  // Check that we got out the same number of values from each bundle.
+  assert(lhsValues.size() == rhsValues.size() &&
+         "connected bundles don't match");
+
+  for (auto tuple : llvm::zip_first(lhsValues, rhsValues)) {
+    Value newLhs = std::get<0>(tuple);
+    Value newRhs = std::get<1>(tuple);
+    builder->create<ConnectOp>(op.getLoc(), newLhs, newRhs);
+  }
+
+  removeOp(op);
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Helpers to manage state.
+//===----------------------------------------------------------------------===//
+
+static DictionaryAttr getArgAttrs(StringAttr nameAttr, StringRef suffix,
+                                  OpBuilder *builder) {
+  SmallString<16> newName(nameAttr.getValue());
+  newName += suffix;
+
+  StringAttr newNameAttr = builder->getStringAttr(newName);
+  Identifier identifier = builder->getIdentifier("firrtl.name");
+  NamedAttribute attr = NamedAttribute(identifier, newNameAttr);
```

Comment:
it'd be slightly simpler to use:

`auto attr = builder->getNamedAttr("firrtl.name", newNameAttr);`